src/runtime/runtime.ts

```diff
@@ -0,0 +1,196 @@
+import { promises as fs } from "fs";
+import BN from "bn.js";
+import * as nearAPI from "near-api-js";
+import { join } from "path";
+import { Account, ContractAccount } from './account'
+import { SandboxServer, getHomeDir } from './server';
+import { debug } from '../utils';
+
+export type TestRunnerFn = (s: Runtime) => Promise<void>;
+
+export interface Config {
+  homeDir: string;
+  port: number;
+  init: boolean;
+  rm: boolean;
+  refDir: string | null;
+  network: 'sandbox' | 'testnet';
+  rootAccountName: string;
+  rpcAddr?: string;
+}
+
+export abstract class Runtime {
+  static async create(config: Partial<Config>): Promise<Runtime> {
+    if (config.network === 'testnet') {
+      return new TestnetRuntime(config)
+    }
+    return new SandboxRuntime(config)
+  }
+
+  abstract get defaultConfig(): Config;
+  abstract run(fn: TestRunnerFn): Promise<Runtime>;
+
+  protected root?: Account;
+  protected near?: nearAPI.Near;
+  protected masterKey?: nearAPI.KeyPair;
+
+  // TODO: should probably be protected
+  config: Config;
+
+  constructor(config: Partial<Config>) {
+    this.config = this.getConfig(config);
+  }
+
+  private getConfig(config: Partial<Config>): Config {
+    return {
+      ...this.defaultConfig,
+      ...config
+    };
+  }
+
+  protected async connect(
+    rpcAddr: string,
+    homeDir: string,
+    init?: boolean
+  ): Promise<void> {
+    const keyFile = require(join(homeDir, "validator_key.json"));
```

Comment:
For testnet should we create a default key here? Or is `validator_key.json` sandbox specific?

Reply:
Perhaps `this.keyFile` should also be abstract.

Reply:
I went ahead and make `getKeyStore` an instance method that subclasses can override.